pulsar-client-cpp/lib/Auth.cc

```diff
@@ -25,82 +26,150 @@
-            ite++) {
-        dlclose(*ite);
+    
+    bool AuthenticationDataProvider::hasDataForHttp() {
+        return false;
     }
-    loadedLibrariesHandles_.clear();
-}
-
-AuthenticationPtr Auth::create(const std::string& dynamicLibPath, const std::string& params) {
-    {
+    
+    std::string AuthenticationDataProvider::getHttpAuthType() {
+        return "none";
+    }
+    
+    std::string AuthenticationDataProvider::getHttpHeaders() {
+        return "none";
+    }
+    
+    bool AuthenticationDataProvider::hasDataFromCommand(){
+        return false;
+    }
+    
+    std::string AuthenticationDataProvider::getCommandData() {
+        return "none";
+    }
+    
+    
+    Authentication::Authentication() {
+    }
+    
+    Authentication::~Authentication() {
+    }
+    
+    class AuthDisabledData : public AuthenticationDataProvider {
+    public:
+        AuthDisabledData(ParamMap& params){
+        }
+    };
+    
+    class AuthDisabled : public Authentication {
+    public:
+        AuthDisabled(AuthenticationDataPtr& authData) {
+            authData_ = authData;
+        }
+        
+        static AuthenticationPtr create(ParamMap& params) {
+            AuthenticationDataPtr authData = AuthenticationDataPtr(new AuthDisabledData(params));
+            return AuthenticationPtr(new AuthDisabled(authData));
+        }
+        
+        const std::string getAuthMethodName() const {
+            return "none";
+        }
+        
+        Result getAuthData(AuthenticationDataPtr& authDataContent) const {
+            authDataContent = authData_;
+            return ResultOk;
+        }
+    private:
+        AuthenticationDataPtr authData_;
+    };
+    
+    
+    AuthenticationPtr Auth::Disabled() {
+        ParamMap params;
+        return AuthDisabled::create(params);
+    }
+    
+    AuthenticationPtr Auth::create(const std::string& dynamicLibPath) {
+        ParamMap params;
+        return Auth::create(dynamicLibPath, params);
+    }
+    
+    boost::mutex mutex;
+    std::vector<void*> Auth::loadedLibrariesHandles_;
+    bool Auth::isShutdownHookRegistered_ = false;
+    
+    void Auth::release_handles() {
         boost::lock_guard<boost::mutex> lock(mutex);
-        if (!Auth::isShutdownHookRegistered_) {
-            atexit(release_handles);
-            Auth::isShutdownHookRegistered_ = true;
+        for (std::vector<void*>::iterator ite = Auth::loadedLibrariesHandles_.begin(); ite != Auth::loadedLibrariesHandles_.end();
+             ite++) {
+            dlclose(*ite);
+        }
+        loadedLibrariesHandles_.clear();
+    }
+    
+    AuthenticationPtr Auth::create(const std::string& dynamicLibPath, const std::string& authParamsString) {
+        ParamMap paramMap;
+        if(!authParamsString.empty()) {
+            std::vector<std::string> params;
+            boost::algorithm::split(params, authParamsString, boost::is_any_of(","));
```

Comment:
It is unlikely that authentication parameter's keys and values contain "\," and "\:" now.
Actually, Java's AuthenticationFactory implementation doesn't worry about it.
https://github.com/yahoo/pulsar/blob/2d75e88269e35ed0e9959c2ad678bd349c605436/pulsar-client/src/main/java/com/yahoo/pulsar/client/api/AuthenticationFactory.java#L42

I think we don't have to consider it right now.
What do you think ?